### PR TITLE
Alerting: Add custom title to DingDing contact point

### PR DESCRIPTION
**Pull request by alexmobo**
**What this PR does / why we need it**:

It allows users to add a custom title for DingDing contact points.
It defaults to title default messages.

### CODE REVIEW

File: pkg/services/ngalert/notifier/channels_config/available_channels.go [hunk 1]
```diff
@@ -132,9 +132,18 @@ func GetAvailableNotifiers() []*NotifierPlugin {
...
+					Element:      ElementTypeInput,
+					InputType:    InputTypeText,
+					Description:  "Templated title of the message",
+					Placeholder:  channels.DefaultMessageTitleEmbed,
+					PropertyName: "title",
+				},
 				{ // New in 8.0.
 					Label:        "Message",
 					Element:      ElementTypeTextArea,
+					Description:  "Templated description of the message",
```
Comment (santihernandezc) [inline pkg/services/ngalert/notifier/channels_config/available_channels.go:146]: This field is pretty self-explanatory. In any case, I think it's better to say something like:

```suggestion
					Description:  "Custom DingDing message. You can use template variables.",
```

**Review by santihernandezc — CHANGES_REQUESTED** [review]
Great! I added two main comments.

File: pkg/services/ngalert/notifier/channels/dingding.go [hunk 2]
```diff
@@ -17,139 +17,148 @@ import (
...
+	Title       string `json:"title,omitempty" yaml:"title,omitempty"`
+	Message     string `json:"message,omitempty" yaml:"message,omitempty"`
 }
 
-func NewDingDingConfig(config *NotificationChannelConfig) (*DingDingConfig, error) {
-	url := config.Settings.Get("url").MustString()
-	if url == "" {
-		return nil, errors.New("could not find url property in settings")
+func buildDingDingSettings(fc FactoryConfig) (dingDingSettings, error) {
+	settings := dingDingSettings{}
```
Comment (santihernandezc) [inline pkg/services/ngalert/notifier/channels/dingding.go:28]: I'd use this syntax for initializing a zero-value struct. Up to you.
```suggestion
	var settings dingDingSettings
```

File: pkg/services/ngalert/notifier/channels/dingding.go [hunk 3]
```diff
@@ -17,139 +17,148 @@ import (
...
 
+func buildDingDingURL(dd *DingDingNotifier) string {
 	q := url.Values{
 		"pc_slide": {"false"},
-		"url":      {ruleURL},
+		"url":      {joinUrlPath(dd.tmpl.ExternalURL.String(), "/alerting/list", dd.log)},
 	}
 
-	// Use special link to auto open the message url outside of Dingding
+	// Use special link to auto open the message url outside Dingding
```
[the remark this answers is not in the document]
Comment (alexmobo) [reply pkg/services/ngalert/notifier/channels/dingding.go:132]: 👍🏽 (I just do what the IDE says 😄)

File: pkg/services/ngalert/notifier/channels/dingding.go [hunk 4]
```diff
@@ -17,139 +17,148 @@ import (
...
+	title := tmpl(dd.settings.Title)
+
+	msgType := tmpl(dd.settings.MessageType)
+	b, err := buildBody(msgUrl, msgType, title, message)
+	if err != nil {
+		return false, err
+	}
+
+	u := tmpl(dd.settings.URL)
+	if tmplErr != nil {
```
[the remark this answers is not in the document]
Comment (alexmobo) [reply pkg/services/ngalert/notifier/channels/dingding.go:113]: Just added an error check

File: pkg/services/ngalert/notifier/channels/template_data.go [hunk 5]
```diff
@@ -149,6 +149,7 @@ func TmplText(ctx context.Context, tmpl *template.Template, alerts []*types.Aler
 	data := ExtendData(promTmplData, l)
 
 	return func(name string) (s string) {
+		// TODO: Make templating function return new error each time instead of passing pointer by parameter
```
Comment (santihernandezc) [inline pkg/services/ngalert/notifier/channels/template_data.go:152]: This is the way that Prometheus handles errors. It was introduced in [this PR](https://github.com/prometheus/alertmanager/pull/1474).
It's kind of an ongoing discussion on [Slack](https://raintank-corp.slack.com/archives/C028MCV4R7C/p1663706834956509). We should come back to this at some point.

**Review by santihernandezc — APPROVED** [review]
LGTM! You can leave or remove the comment I mention in this review, your choice.

**Pull request merged**